libherokubuildpack/src/command.rs

```diff
@@ -0,0 +1,185 @@
+use crate::write::tee;
+use crossbeam::thread::ScopedJoinHandle;
+use std::io::Write;
+use std::{io, process, thread};
+use std::{mem, panic};
+
+/// Extension trait for [`process::Command`] that adds functions for use within buildpacks.
+pub trait CommandExt {
+    /// Spawns the command process and sends the output of stdout and stderr to the given writers.
+    ///
+    /// This allows for additional flexibility when dealing with these output streams compared to
+    /// functionality that the stock [`process::Command`] provides. See the [`write`](crate::write)
+    /// module for [`std::io::Write`] implementations designed for common buildpack tasks.
+    ///
+    /// This function will redirect the output unbuffered and in parallel for both streams. This
+    /// means that it can be used to output data from these streams while the command is running,
+    /// providing a live view into the process' output. This function will block until both streams
+    /// have been closed.
+    ///
+    /// # Example:
+    /// ```no_run
+    /// use libherokubuildpack::command::CommandExt;
+    /// use libherokubuildpack::write::tee;
+    /// use std::fs;
+    /// use std::process::Command;
+    ///
+    /// let logfile = fs::File::open("log.txt").unwrap();
+    /// let exit_status = Command::new("date")
+    ///     .spawn_and_write_streams(tee(std::io::stdout(), &logfile), std::io::stderr())
+    ///     .and_then(|mut child| child.wait())
+    ///     .unwrap();
+    /// ```
+    fn spawn_and_write_streams<OW: Write + Send, EW: Write + Send>(
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Child>;
+
+    /// Spawns the command process and sends the output of stdout and stderr to the given writers.
+    ///
+    /// In addition to what [`spawn_and_write_streams`](Self::spawn_and_write_streams) does, this
+    /// function captures stdout and stderr as `Vec<u8>` and returns them after waiting for the
+    /// process to finish. This function is meant as a drop-in replacement for existing
+    /// `Command:output` calls.
+    ///
+    /// # Example:
+    /// ```no_run
+    /// use libherokubuildpack::command::CommandExt;
+    /// use libherokubuildpack::write::tee;
+    /// use std::fs;
+    /// use std::process::Command;
+    ///
+    /// let logfile = fs::File::open("log.txt").unwrap();
+    /// let output = Command::new("date")
+    ///     .output_and_write_streams(tee(std::io::stdout(), &logfile), std::io::stderr())
+    ///     .unwrap();
+    ///
+    /// // Return value can be used as with Command::output, but the streams will also be written to
+    /// // the given writers.
+    /// println!(
+    ///     "Process exited with {}, stdout: {:?}, stderr: {:?}",
+    ///     output.status, output.stdout, output.stderr
+    /// );
+    /// ```
+    fn output_and_write_streams<OW: Write + Send, EW: Write + Send>(
```

Comment:
I don't love these method names. I know they match Command's but I find them confusing. 

Looking at the code both `output` and  `spawn` seem like generic actions rather than behaviors.

Would you be open to `capture_output_and_write_streams` or `capture_and_write_streams`. Maybe just `capture_and_stream`.

Reply:
Hm, I'm quite happy with the naming of these - especially compared to earlier iterations I had. 😂 

I think it's beneficial to name them after the `Command` functions they emulate (they are drop-in replacements). I think of these functions and an additional "tap" for the output streams - in the form of the `Write` interface. 

Do you think that `Command::spawn` and `Command::output` should be renamed to `Command::capture` and `Command::capture_output` too? I think at least `capture` is a bit misleading. In any case, I don't see the benefit of inventing another name for the same thing that users then have to lean in addition - no matter how "bad" the original naming is.

Reply:
TLDR; I agree but it took awhile to get there

> Do you think that Command::spawn and Command::output should be renamed to Command::capture and Command::capture_output too? 

I care about behaviors `Command::capture_output` would be better than `Command::output` imho. I guess you could advocate for `Command::output` being named because it GIVES you the output. But if that's the case `Command::spawn` makes no sense because it does not GIVE you the `spawn` (what does that even mean?). It gives you a child which I guess is a spawned process https://doc.rust-lang.org/std/process/struct.Child.html. So, meh...maybe that makes sense. But I've been using `Command` for a long time and it took me until now to make that connection. 

I think that it __technically__ makes sense, but only when you squint. I have a comparison to these functions: 

https://github.com/heroku/buildpacks-ruby/blob/schneems/fresh-start-rust/commons/src/in_app_dir_cache.rs

While technically `copy_app_path_to_cache` and `move_app_path_to_cache` are correctly named but the subtle difference is easy to miss, so I renamed one of them `destructive_move_app_path_to_cache` so it is PAINFULLY obvious what you're getting it when reading it.

So i think to sum up my argument: `output` and `spawn` might be technically correct and an okay name, but I don't think they communicate behavior or intent to someone coming from a different language or context. I feel that command is actually too low level of a primitive (hence why I wrote my own wrapper struct that none of you like). 

> I think at least capture is a bit misleading.

Can you say more here?

> I don't see the benefit of inventing another name for the same thing that users then have to learn in addition - no matter how "bad" the original naming is.

I'm thinking about it and using my own arguments...I agree. The difference between capturing and not capturing would be detected at code time as the developer would realize they can't access the stdout/stderr. Versus in the stream/no-stream case the difference to customers is huge. However in both our methods we are streaming and it explicitly says so in the method name `*_and_write_streams`

I'm leaving this long comment, because it's how I derived this conclusion and want to leave it as a note to myself.

Reply:
I'm marking this is "resolved", LMK if I misunderstood your comment and feel free to re-open! :)